### PR TITLE
Introduce Action Group resource of Azure Monitor

### LINKED ISSUE #1252: New Resource: Azure Monitor Signal
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Terraform supports creating metrics-based alerts with `azurerm_metric_alertrule` but I don't see any support for creating newer "signal" logic alerts.  For example, from the Azure Portal, I can create an alert to notify me on Slack when a new App Service state changes (start/stop etc).

More information on the new alerts:
https://docs.microsoft.com/en-us/azure/monitoring-and-diagnostics/monitoring-overview-unified-alerts

### New or Affected Resource(s)

A new resource for managing the the newer qalerts experience in Azure Monitor.

### PR DESCRIPTION
**Pull request by JunyiYi**
In this PR, I introduced a new simple action group resource of Azure monitor.
I enabled three receivers (Email, SMS and Webhook) which should cover most of the use cases.

This PR is a **first step** of #1252 , because the resource (signal based alert) proposed in it relies on Action Group.

(fixes #1252)

### CODE REVIEW

File: azurerm/resource_arm_action_group.go [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+		Delete: resourceArmActionGroupDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
```
Comment (jeffreyCline) [inline azurerm/resource_arm_action_group.go]: Could we add some validation here, like no empty values?

File: azurerm/resource_arm_action_group.go [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+			},
+
+			"location":            locationSchema(),
+			"resource_group_name": resourceGroupNameSchema(),
```
Comment (jeffreyCline) [inline azurerm/resource_arm_action_group.go]: This is really minor, but can we get a space between these too?

File: azurerm/resource_arm_action_group.go [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+			result = append(result, val)
+		}
+	}
+	return result
+}
+
+func expandActionGroupSmsReceiver(d *schema.ResourceData) *[]insights.SmsReceiver {
+	v, ok := d.GetOk("sms_receiver")
+	if !ok {
+		return nil
```
Comment (jeffreyCline) [inline azurerm/resource_arm_action_group.go]: Unless there is a reason to return nil here, can we return an empty array as above?

File: azurerm/resource_arm_action_group.go [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+			result = append(result, val)
+		}
+	}
+	return result
+}
+
+func expandActionGroupWebHookReceiver(d *schema.ResourceData) *[]insights.WebhookReceiver {
+	v, ok := d.GetOk("webhook_receiver")
+	if !ok {
+		return nil
```
Comment (jeffreyCline) [inline azurerm/resource_arm_action_group.go]: Unless there is a reason to return nil here, can we return an empty array as above?

File: azurerm/import_arm_action_group_test.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
+package azurerm
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+)
+
+func TestAccAzureRMActionGroup_importBasic(t *testing.T) {
```
Comment (jeffreyCline) [inline azurerm/import_arm_action_group_test.go]: Minor, you could move the import tests in with the regular tests if you would like. It looks like this is the direction that other providers are going, so we may want to do this as well.

Comment (JunyiYi) [reply azurerm/import_arm_action_group_test.go]: Moved to `resource_arm_action_group_test.go`

File: azurerm/resource_arm_action_group.go (hunk 1 above)
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: Done.

File: azurerm/resource_arm_action_group.go (hunk 2 above)
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: Sure.

File: azurerm/resource_arm_action_group.go [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"location":            locationSchema(),
+			"resource_group_name": resourceGroupNameSchema(),
+
+			"short_name": {
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: Nice catch 👍

File: azurerm/resource_arm_action_group.go [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+			},
+
+			"email_receiver": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Required: true,
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: Tried with Azure APIs to see which fields should be non-empty.

File: azurerm/resource_arm_action_group.go [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+	shortName := d.Get("short_name").(string)
+	enabled := d.Get("enabled").(bool)
+
+	tags := d.Get("tags").(map[string]interface{})
+	expandedTags := expandTags(tags)
+
+	parameters := insights.ActionGroupResource{
+		Location: utils.String(location),
+		ActionGroup: &insights.ActionGroup{
+			GroupShortName:   utils.String(shortName),
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: Resolved.

File: azurerm/resource_arm_action_group.go [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+		}
+		return fmt.Errorf("Error deleting action group %s (resource group %s): %+v", name, resGroup, err)
+	}
+
+	return nil
+}
+
+func expandActionGroupEmailReceiver(d *schema.ResourceData) *[]insights.EmailReceiver {
+	v, ok := d.GetOk("email_receiver")
+	if !ok {
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: Absolutely.

File: azurerm/resource_arm_action_group.go [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+			Name:         utils.String(val["name"].(string)),
+			EmailAddress: utils.String(val["email_address"].(string)),
+		}
+
+		receivers = append(receivers, receiver)
+	}
+	return &receivers
+}
+
+func flattenActionGroupEmailReceiver(receivers *[]insights.EmailReceiver) []interface{} {
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: Definitely.

File: azurerm/resource_arm_action_group.go (hunk 3 above)
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: Done.

File: azurerm/resource_arm_action_group.go (hunk 4 above)
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: Done.

File: azurerm/resource_arm_action_group_test.go [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+					resource.TestCheckResourceAttr(resourceName, "sms_receiver.0.phone_number", "1231231234"),
+					resource.TestCheckResourceAttr(resourceName, "webhook_receiver.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "webhook_receiver.0.service_uri", "http://example.com/alert"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAzureRMActionGroup_importBasic(t *testing.T) {
```
Comment (jeffreyCline) [inline azurerm/resource_arm_action_group_test.go]: Can we delete the import function and add the import test as an additional test step in the existing resource test? Sorta like this (do this for all import tests in the PR)...

```
Steps: []resource.TestStep{ 
	{ 
		Config: config, 
		Check: resource.ComposeTestCheckFunc( 
			testCheckAzureRMActionGroupExists(resourceName), 
			resource.TestCheckResourceAttr(resourceName, "enabled", "true"), 
			resource.TestCheckResourceAttr(resourceName, "email_receiver.#", "2"), 
			resource.TestCheckResourceAttr(resourceName, "email_receiver.0.email_address", "admin@contoso.com"), 
			resource.TestCheckResourceAttr(resourceName, "email_receiver.1.email_address", "devops@contoso.com"), 
			resource.TestCheckResourceAttr(resourceName, "sms_receiver.#", "1"), 
			resource.TestCheckResourceAttr(resourceName, "sms_receiver.0.country_code", "1"), 
			resource.TestCheckResourceAttr(resourceName, "sms_receiver.0.phone_number", "1231231234"), 
			resource.TestCheckResourceAttr(resourceName, "webhook_receiver.#", "1"), 
			resource.TestCheckResourceAttr(resourceName, "webhook_receiver.0.service_uri", "http://example.com/alert"), 
		), 
	},
	{ 
		ResourceName:      resourceName, 
		ImportState:       true, 
		ImportStateVerify: true, 
	}, 

```

Comment (JunyiYi) [reply azurerm/resource_arm_action_group_test.go]: I've changed the test cases.

File: website/docs/r/action_group.html.markdown [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the Action Group. Changing this forces a new resource to be created.
+* `location` - (Required) The location of this Action Group. The only possible value is `Global`.
```
Comment (tombuildsstuff) [inline website/docs/r/action_group.html.markdown]: if the only possible value is `Global` then this field should be hard-coded and shouldn't be available for users to set

File: website/docs/r/action_group.html.markdown [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+* `resource_group_name` - (Required) The name of the resource group in which to create the Action Group instance.
+* `short_name` - (Required) The short name of the action group. This will be used in SMS messages.
+* `enabled` - (Optional) Whether this action group is enabled. If an action group is not enabled, then none of its receivers will receive communications. Defaults to `true`.
+* `email_receiver` - (Optional) The list of `email_receiver` blocks as defined below that are part of this action group.
+* `sms_receiver` - (Optional) The list of `sms_receiver` blocks as defined below that are part of this action group.
+* `webhook_receiver` - (Optional) The list of `webhook_receiver` blocks as defined below that are part of this action group.
```
Comment (tombuildsstuff) [inline website/docs/r/action_group.html.markdown]: All three of these should be updated to match the other resources:

```
* `email_receiver ` - (Optional) One or more `email_receiver ` blocks as defined below.
* `sms_receiver ` - (Optional) One or more `sms_receiver ` blocks as defined below.
* `webhook_receiver` - (Optional) One or more `webhook_receiver` blocks as defined below.
```

File: website/docs/r/action_group.html.markdown [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+`webhook_receiver` supports the following:
+
+* `name` - (Required) The name of the webhook receiver. Names must be unique (case-insensitive) across all receivers within an action group.
+* `service_uri` - (Required) The URI where webhooks should be sent.
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `id` - The Route ID.
```
Comment (tombuildsstuff) [inline website/docs/r/action_group.html.markdown]: presumably this should be "The ID of the Action Group"?

File: azurerm/resource_arm_action_group_test.go [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+					testCheckAzureRMActionGroupExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "enabled", "true"),
+					resource.TestCheckResourceAttr(resourceName, "email_receiver.#", "2"),
+					resource.TestCheckResourceAttr(resourceName, "email_receiver.0.email_address", "admin@contoso.com"),
+					resource.TestCheckResourceAttr(resourceName, "email_receiver.1.email_address", "devops@contoso.com"),
+					resource.TestCheckResourceAttr(resourceName, "sms_receiver.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "sms_receiver.0.country_code", "1"),
+					resource.TestCheckResourceAttr(resourceName, "sms_receiver.0.phone_number", "1231231234"),
+					resource.TestCheckResourceAttr(resourceName, "webhook_receiver.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "webhook_receiver.0.service_uri", "http://example.com/alert"),
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_action_group_test.go]: can we add some tests to go from empty -> basic and ensure these fields are set?

File: azurerm/resource_arm_action_group_test.go [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMActionGroupExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "enabled", "false"),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_action_group_test.go]: in general we don't test importing the updates, but rather full configurations. in this case can we split this test into two: one to test with it in the disabled state, and one to ensure it can go from disabled -> enabled -> disabled

File: azurerm/resource_arm_action_group.go [hunk 17]
```diff
@@ -0,0 +1,313 @@
...
+	return nil
+}
+
+func resourceArmActionGroupDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).actionGroupsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return fmt.Errorf("Error parsing action group resource ID \"%s\" during delete: %+v", d.Id(), err)
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_action_group.go]: we don't need to wrap this error message - so can we just return it directly?

File: azurerm/resource_arm_action_group.go [hunk 18]
```diff
@@ -0,0 +1,313 @@
...
+	}
+
+	d.Set("name", name)
+	d.Set("resource_group_name", resGroup)
+	if location := resp.Location; location != nil {
+		d.Set("location", azureRMNormalizeLocation(*location))
+	}
+
+	d.Set("short_name", *resp.GroupShortName)
+	d.Set("enabled", *resp.Enabled)
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: It is a shortcut in go instead of accessing `.Properties`.

File: azurerm/resource_arm_action_group.go [hunk 19]
```diff
@@ -0,0 +1,313 @@
...
+	return resourceArmActionGroupRead(d, meta)
+}
+
+func resourceArmActionGroupRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).actionGroupsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return fmt.Errorf("Error parsing action group resource ID \"%s\" during get: %+v", d.Id(), err)
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: That's my personal style. I don't think it really matters.

File: azurerm/resource_arm_action_group.go [hunk 20]
```diff
@@ -0,0 +1,313 @@
...
+	enabled := d.Get("enabled").(bool)
+
+	tags := d.Get("tags").(map[string]interface{})
+	expandedTags := expandTags(tags)
+
+	parameters := insights.ActionGroupResource{
+		Location: &location,
+		ActionGroup: &insights.ActionGroup{
+			GroupShortName: &shortName,
+			Enabled:        &enabled,
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: According to @katbyte (@jeffreyCline could confirm that) , the new style is to use `addressof` instead of utilities.

File: azurerm/resource_arm_action_group.go [hunk 21]
```diff
@@ -0,0 +1,313 @@
...
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validation.NoZeroValues,
+						},
+						"service_uri": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validation.NoZeroValues,
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: We won't because Azure won't. We can pass anything to Azure. And if in the future Azure does, API call will return the error message.

File: azurerm/resource_arm_action_group.go [hunk 22]
```diff
@@ -0,0 +1,313 @@
...
+	return &receivers
+}
+
+func flattenActionGroupEmailReceiver(receivers *[]insights.EmailReceiver) []interface{} {
+	result := make([]interface{}, 0)
+	if receivers != nil {
+		for _, receiver := range *receivers {
+			val := make(map[string]interface{}, 0)
+			val["name"] = *receiver.Name
+			val["email_address"] = *receiver.EmailAddress
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: Azure won't allow nil values for these fields.

File: azurerm/resource_arm_action_group.go [hunk 23]
```diff
@@ -0,0 +1,313 @@
...
+}
+
+func flattenActionGroupSmsReceiver(receivers *[]insights.SmsReceiver) []interface{} {
+	result := make([]interface{}, 0)
+	if receivers != nil {
+		for _, receiver := range *receivers {
+			val := make(map[string]interface{}, 0)
+			val["name"] = *receiver.Name
+			val["country_code"] = *receiver.CountryCode
+			val["phone_number"] = *receiver.PhoneNumber
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: Azure won't allow nil values for these fields.

File: azurerm/resource_arm_action_group.go [hunk 24]
```diff
@@ -0,0 +1,313 @@
...
+	return result
+}
+
+func flattenActionGroupWebHookReceiver(receivers *[]insights.WebhookReceiver) []interface{} {
+	result := make([]interface{}, 0)
+	if receivers != nil {
+		for _, receiver := range *receivers {
+			val := make(map[string]interface{}, 0)
+			val["name"] = *receiver.Name
+			val["service_uri"] = *receiver.ServiceURI
```
[the remark this answers is not in the document]
Comment (JunyiYi) [reply azurerm/resource_arm_action_group.go]: Azure won't allow nil values for these fields.

File: azurerm/resource_arm_action_group.go (hunk 17 above)
Comment (tombuildsstuff) [reply azurerm/resource_arm_action_group.go]: we consistently across the codebase don't wrap these error messages since they're an internal error users aren't going to be able to work around (vs a temporary issue in the Azure API in CRUD/Polling where they may be able to self-diagnose) - as such ID parsing error don't need wrapping.

File: website/docs/r/action_group.html.markdown (hunk 13 above)
Comment (tombuildsstuff) [reply website/docs/r/action_group.html.markdown]: Newer resources / changes use this format, there's a PR in-flight to migrate others over:

- https://www.terraform.io/docs/providers/azurerm/r/virtual_machine.html#storage_data_disk
- https://www.terraform.io/docs/providers/azurerm/r/kubernetes_cluster.html#agent_pool_profile
-
